Use Apache Mina, remove trilead for ssh connectivity

olamy opened the pull request:
## Use Apache Mina, remove trilead for ssh connectivity

- Replace trilead-api with Apache Mina ssh

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-client-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] I have interactively tested my changes

## Types of changes

What types of changes does your code introduce?

- [x] New feature (non-breaking change which adds functionality)

jtnord commented on pom.xml:
```diff
@@ -202,15 +202,17 @@
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>ssh-credentials</artifactId>
+      <exclusions>
+        <exclusion>
+          <groupId>org.jenkins-ci.plugins</groupId>
+          <artifactId>trilead-api</artifactId>
+        </exclusion>
+      </exclusions>
```
should be an optional dependency so it should not be transitive here?

jtnord commented on src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java:
```diff
@@ -166,15 +179,99 @@ public class JGitAPIImpl extends LegacyCompatibleGitAPIImpl {
+                        .toPath());
+                Files.createFile(SshHostKeyVerificationStrategy.JGIT_KNOWN_HOSTS_FILE.toPath());
+            } catch (IOException e) {
+                LOGGER.log(Level.SEVERE, "cannot create knowhosts file", e);
```
```suggestion
                LOGGER.log(Level.SEVERE, "cannot create known hosts file", e);
```

is this not run on the controller - as such won't this create / manipulate the wrong file?  A build of one project configured in a specific way would impact all projects even ones run on different hosts.

jtnord commented on src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java:
```diff
@@ -166,15 +179,99 @@ public class JGitAPIImpl extends LegacyCompatibleGitAPIImpl {
+                LOGGER.log(Level.SEVERE, "cannot create knowhosts file", e);
+            }
+        }
+        return new SshdSessionFactory() {
```
the factory will have a reference to the outer object.   does this cause issues if used on an agent?

olamy replied:
nope

jtnord commented on src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java:
```diff
@@ -166,15 +179,99 @@ public class JGitAPIImpl extends LegacyCompatibleGitAPIImpl {
+                        .findFirst()
+                        .orElse(null);
+                if (sshUserPrivateKey != null) {
+                    try {
+                        // be sure parent directories are here
+                        Files.createDirectories(sshDir.getAbsoluteFile().toPath());
+                        tmpKey = Files.createTempFile(sshDir.getAbsoluteFile().toPath(), "key", ".priv");
+                        tmpKey.toFile().deleteOnExit();
+                        Files.write(tmpKey, sshUserPrivateKey.getPrivateKeys());
+                        return Collections.singletonList(tmpKey);
```
I think should be using a `workspace@tmp` or the like?  
Files.createDirectories will in any case, create the `.ssh` directory with incorrect perms which can/will break any use of the native ssh on the host.

jtnord commented on src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java:
```diff
@@ -166,15 +179,99 @@ public class JGitAPIImpl extends LegacyCompatibleGitAPIImpl {
+
+            @Override
+            protected String getDefaultPreferredAuthentications() {
+                // TODO do we really need/want password?
+                return "publickey,password";
```
likely would depend on the passed in credential, but possibly yes

jtnord commented on src/main/java/org/jenkinsci/plugins/gitclient/verifier/AbstractJGitHostKeyVerifier.java:
```diff
@@ -1,64 +1,20 @@
 import org.jenkinsci.remoting.SerializableOnlyOverRemoting;
 
-public abstract class AbstractJGitHostKeyVerifier implements ServerHostKeyVerifier, SerializableOnlyOverRemoting {
+public abstract class AbstractJGitHostKeyVerifier implements SerializableOnlyOverRemoting {
```
what am I missing that this does not implement [org.apache.sshd.client.keyverifier.ServerKeyVerifier](https://javadoc.io/doc/org.apache.sshd/sshd-core/latest/org/apache/sshd/client/keyverifier/ServerKeyVerifier.html)

jtnord commented on src/main/java/org/jenkinsci/plugins/gitclient/verifier/AbstractJGitHostKeyVerifier.java:
```diff
@@ -1,64 +1,20 @@
-    KnownHosts getKnownHosts() {
-        return knownHosts;
-    }
+    public abstract OpenSshConfigFile.HostEntry customizeHostEntry(OpenSshConfigFile.HostEntry hostEntry);
```
unclear why we are customizing the known hosts, rather than using https://javadoc.io/static/org.apache.sshd/sshd-core/2.12.1/org/apache/sshd/client/keyverifier/ServerKeyVerifier.html#verifyServerKey(org.apache.sshd.client.session.ClientSession,java.net.SocketAddress,java.security.PublicKey) ?

jtnord commented on src/main/java/org/jenkinsci/plugins/gitclient/verifier/AcceptFirstConnectionVerifier.java:
```diff
@@ -27,95 +23,32 @@ public AbstractCliGitHostKeyVerifier forCliGit(TaskListener listener) {
-                    isValid = true;
-                } else if (KnownHosts.HOSTKEY_IS_NEW == hostPortResult) {
-                    writeToFile(knownHostsFile, hostnamePort, serverHostKeyAlgorithm, serverHostKey);
-                    isValid = true;
```
why not just use [DefaultKnownHostsServerKeyVerifier](https://javadoc.io/static/org.apache.sshd/sshd-core/2.12.1/org/apache/sshd/client/keyverifier/DefaultKnownHostsServerKeyVerifier.html#%3Cinit%3E(org.apache.sshd.client.keyverifier.ServerKeyVerifier,boolean))([ServerKeyVerifier](https://javadoc.io/static/org.apache.sshd/sshd-core/2.12.1/org/apache/sshd/client/keyverifier/ServerKeyVerifier.html) delegate, boolean strict) strict set to false?

olamy replied:
neither https://javadoc.io/doc/org.eclipse.jgit/org.eclipse.jgit.ssh.apache/latest/org.eclipse.jgit.ssh.apache/org/eclipse/jgit/transport/sshd/SshdSessionFactory.html or https://javadoc.io/doc/org.eclipse.jgit/org.eclipse.jgit.ssh.apache/latest/org.eclipse.jgit.ssh.apache/org/eclipse/jgit/transport/sshd/SshdSessionFactoryBuilder.html offer the possibility to configure this.

I don't like much the idea of having to maintain a re-implementation of this code https://github.com/eclipse-jgit/jgit/blob/5c1c006307cac302b07d42ed3172a3cb3e630dcc/org.eclipse.jgit.ssh.apache/src/org/eclipse/jgit/transport/sshd/SshdSessionFactory.java#L191

jtnord replied:
what about `JGitSshClient.setServerKeyVerifier()` ?   do we need to be configuring the `SshdSessionFactory` directly rather than letting jgitClient do what we want it do with the twaeaks for hosts/credentials?

olamy replied:
> what about `JGitSshClient.setServerKeyVerifier()` ? do we need to be configuring the `SshdSessionFactory` directly rather than letting jgitClient do what we want it do with the twaeaks for hosts/credentials?

no access to this (it's all static and private) and no configurable (except as said previously having our own implementation `SshdSessionFactory`) FYI this `SshdSessionFactory` creating  `JGitSshClient` (I'm not quite sure what you mean with `jgitClient`)
Again everything is using some statics method without much configuration option such: 
- SshSessionFactory.setInstance
- HttpTransport.setConnectionFactory

look at the branch locally and look at what are the possible methods. configuration are really limited and based on some static methods and system property.

jtnord replied:
```java
class MySshdSessionFactory extends SshdSessionFactory {

  @override
  public SshdSession getSession(URIish uri,
			CredentialsProvider credentialsProvider, FS fs, int tms)
			throws TransportException {
      SshClient client = super.getSession(uri, credentialsProvider, ds, tms);
      client.setServerKeyVerifier(....)    
      ...
   }
}
```

jtnord replied:
```java
	@Override
	public SshdSession getSession(URIish uri,
			CredentialsProvider credentialsProvider, FS fs, int tms)
			throws TransportException {
		SshdSession session = null;
		try {
			session = new SshdSession(uri, () -> {
				File home = getHomeDirectory();
				if (home == null) {
					// Always use the detected filesystem for the user home!
					// It makes no sense to have different "user home"
					// directories depending on what file system a repository
					// is.
					home = FS.DETECTED.userHome();
				}
				File sshDir = getSshDirectory();
				if (sshDir == null) {
					sshDir = new File(home, SshConstants.SSH_DIR);
				}
				HostConfigEntryResolver configFile = getHostConfigEntryResolver(
						home, sshDir);
				KeyIdentityProvider defaultKeysProvider = toKeyIdentityProvider(
						getDefaultKeys(sshDir));
				Supplier<KeyPasswordProvider> keyPasswordProvider = () -> createKeyPasswordProvider(
						credentialsProvider);
				SshClient client = ClientBuilder.builder()
						.factory(JGitSshClient::new)
						.filePasswordProvider(createFilePasswordProvider(
								keyPasswordProvider))
						.hostConfigEntryResolver(configFile) // XXX customize me
						.serverKeyVerifier(new JGitServerKeyVerifier(
								getServerKeyDatabase(home, sshDir))) // XXX customize me
						.signatureFactories(getSignatureFactories())
						.compressionFactories(
								new ArrayList<>(BuiltinCompressions.VALUES))
						.build();
				client.setUserInteraction(
						new JGitUserInteraction(credentialsProvider));
				client.setUserAuthFactories(getUserAuthFactories());
				client.setKeyIdentityProvider(defaultKeysProvider);
				ConnectorFactory connectors = getConnectorFactory();
				if (connectors != null) {
					client.setAgentFactory(
							new JGitSshAgentFactory(connectors, home));
				}
				// JGit-specific things:
				JGitSshClient jgitClient = (JGitSshClient) client;
				jgitClient.setKeyCache(getKeyCache());
				jgitClient.setCredentialsProvider(credentialsProvider);
				jgitClient.setProxyDatabase(proxies);
				jgitClient.setKeyPasswordProviderFactory(keyPasswordProvider);
				String defaultAuths = getDefaultPreferredAuthentications();
				if (defaultAuths != null) {
					jgitClient.setAttribute(
							JGitSshClient.PREFERRED_AUTHENTICATIONS,
							defaultAuths);
				}
				if (home != null) {
					try {
						jgitClient.setAttribute(JGitSshClient.HOME_DIRECTORY,
								home.getAbsoluteFile().toPath());
					} catch (SecurityException | InvalidPathException e) {
					// Ignore
					}
				}
				// Other things?
				return client;
			});
			session.addCloseListener(s -> unregister(s));
			register(session);
			session.connect(Duration.ofMillis(tms));
			return session;
		} catch (Exception e) {
			unregister(session);
			if (e instanceof TransportException) {
				throw (TransportException) e;
			}
			throw new TransportException(uri, e.getMessage(), e);
		}
	}
```

olamy replied:
yup I wanted to avoid this copy of so much internal code. 
but at least it's still less code duplication than an unmaintained trilead fork :)

jtnord replied:
If this is a desirable way to customize we could send a PR to Apache mina to make the "create the jgitsshclient" a method of the factory rather so it can more easily be overridden to remove the copy paste and leave only the things you want to override.

olamy replied:
we might need first PR to `org.eclipse.jgit.transport.sshd.SshdSessionFactoryBuilder`.

olamy replied:
well this will be really crappy.... 
as the method `getSession` to override is using a lot of private methods (to duplicate) 
SshdSession (the one we need to create an instance) is package-private, so it cannot be used. except using the same package.
So turns into:
```
java.lang.SecurityException: class "org.eclipse.jgit.transport.sshd.JenkinsSshdSessionFactory"'s signer ```
information does not match signer information of other classes in the same package
```

tbh that's very ugly..... (probably worst than what is done now)

jtnord commented on src/main/java/org/jenkinsci/plugins/gitclient/verifier/KnownHostsFileVerifier.java:
```diff
@@ -32,51 +32,29 @@ public AbstractCliGitHostKeyVerifier forCliGit(TaskListener listener) {
-            });
-            return verifyServerHostKey(
-                    listener, getKnownHosts(), hostname, port, serverHostKeyAlgorithm, serverHostKey);
+        public OpenSshConfigFile.HostEntry customizeHostEntry(OpenSshConfigFile.HostEntry hostEntry) {
+            hostEntry.setValue(SshConstants.STRICT_HOST_KEY_CHECKING, SshConstants.YES);
+            return hostEntry;
```
confused why we are working at a low level when higher primatives exist that can seemingly hide this complexity?

olamy replied:
https://github.com/jenkinsci/git-client-plugin/pull/1127#discussion_r1566668664

jtnord commented on src/main/java/org/jenkinsci/plugins/gitclient/verifier/ManuallyProvidedKeyVerifier.java:
```diff
@@ -34,53 +34,60 @@ public AbstractCliGitHostKeyVerifier forCliGit(TaskListener listener) {
-        public ManuallyProvidedKeyJGitHostKeyVerifier(TaskListener listener, KnownHosts knownHosts) {
-            super(knownHosts);
-            this.listener = listener;
+        public ManuallyProvidedKeyJGitHostKeyVerifier(TaskListener listener, String approvedHostKeys) {
```
see https://javadoc.io/static/org.apache.sshd/sshd-core/2.12.1/org/apache/sshd/client/keyverifier/RequiredServerKeyVerifier.html ?

olamy replied:
https://github.com/jenkinsci/git-client-plugin/pull/1127#discussion_r1566668664

jtnord commented on src/main/java/org/jenkinsci/plugins/gitclient/verifier/NoHostKeyVerifier.java:
```diff
@@ -18,26 +16,11 @@ public AbstractCliGitHostKeyVerifier forCliGit(TaskListener listener) {
-                            Base64.getEncoder().encodeToString(serverHostKey)
-                        });
-                return true;
+            public OpenSshConfigFile.HostEntry customizeHostEntry(OpenSshConfigFile.HostEntry hostEntry) {
```
https://javadoc.io/static/org.apache.sshd/sshd-core/2.12.1/org/apache/sshd/client/keyverifier/AcceptAllServerKeyVerifier.html ?

olamy replied:
https://github.com/jenkinsci/git-client-plugin/pull/1127#discussion_r1566668664

jtnord commented on src/main/java/org/jenkinsci/plugins/gitclient/verifier/SshHostKeyVerificationStrategy.java:
```diff
@@ -12,7 +12,7 @@ public abstract class SshHostKeyVerificationStrategy<T extends HostKeyVerifierFa
 
     public static final String KNOWN_HOSTS_DEFAULT =
-            Paths.get(System.getProperty("user.home"), ".ssh", "known_hosts").toString();
+            Paths.get("target", ".ssh", "known_hosts").toString();
```
? production code not test, so the use of `target` here seems incorrect?

olamy replied:
because it's still marked as WIP and draft. While working on this I don't want to break my `~/.ssh content`

olamy commented on src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java:
```diff
@@ -163,16 +184,91 @@ public class JGitAPIImpl extends LegacyCompatibleGitAPIImpl {
+            protected Iterable<KeyPair> getDefaultKeys(File sshDir) {
+                SmartCredentialsProvider smartCredentialsProvider = JGitAPIImpl.getProvider();
+                // credentials have been added for the target repo only so we assume we have only the good one here
+                SSHUserPrivateKey sshUserPrivateKey = smartCredentialsProvider.getCredentials().values().stream()
```
ideally, here we should have access to the target URL to use only this credential ()

jtnord commented on pom.xml:
```diff
@@ -103,6 +103,19 @@
       <artifactId>mina-sshd-api-core</artifactId>
     </dependency>
+    <!-- this could be optional? -->
+    <dependency>
+      <groupId>net.i2p.crypto</groupId>
+      <artifactId>eddsa</artifactId>
+      <version>0.3.0</version>
+      <optional>true</optional>
+    </dependency>
+    <dependency>
+      <groupId>org.bouncycastle</groupId>
+      <artifactId>bcprov-jdk18on</artifactId>
+      <version>1.77</version>
+      <optional>true</optional>
+    </dependency>
```
```suggestion
```

https://github.com/jenkinsci/mina-sshd-api-plugin/pull/98  these are a dependency of mina not the git plugin.

olamy replied:
done https://github.com/jenkinsci/git-client-plugin/pull/1127/commits/1b368b0e5295076894b6719c582e6302275032f4

olamy commented:
currently 3 new tests are failing because using testcontainers and need some changes in Jenkinsfile but it's not taken into account.

MarkEWaite commented:
Thanks very much @olamy !

I've pushed the Jenkinsfile change to the master branch so that the plugin can use testcontainers.

jtnord reviewed (commented):
I have 3 failures locally

```
[ERROR]   AuthzTest.httpWithPassword:93->testRepo:125 » IllegalArgument Invalid value: gpg.format=ssh
[ERROR]   AuthzTest.sshKeyAuth:57->testRepo:125 » IllegalArgument Invalid value: gpg.format=ssh
[ERROR]   AuthzTest.sshWithPassword:76->testRepo:125 » IllegalArgument Invalid value: gpg.format=ssh
```

```
[ERROR] org.jenkinsci.plugins.gitclient.jgit.AuthzTest.httpWithPassword -- Time elapsed: 7.125 s <<< ERROR!
java.lang.IllegalArgumentException: Invalid value: gpg.format=ssh
        at org.eclipse.jgit.lib.DefaultTypedConfigGetter.getEnum(DefaultTypedConfigGetter.java:103)
        at org.eclipse.jgit.lib.Config.getEnum(Config.java:454)
        at org.eclipse.jgit.lib.GpgConfig.<init>(GpgConfig.java:86)
        at org.eclipse.jgit.api.CommitCommand.processOptions(CommitCommand.java:662)
        at org.eclipse.jgit.api.CommitCommand.call(CommitCommand.java:189)
        at org.jenkinsci.plugins.gitclient.JGitAPIImpl.commit(JGitAPIImpl.java:555)
        at org.jenkinsci.plugins.gitclient.jgit.AuthzTest.testRepo(AuthzTest.java:125)
        at org.jenkinsci.plugins.gitclient.jgit.AuthzTest.httpWithPassword(AuthzTest.java:93)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
```

see https://github.com/jenkinsci/git-client-plugin/pull/1124 for previous bad assumptions and how they where fixed

olamy commented on pom.xml:
```diff
@@ -66,6 +66,7 @@
     <!-- https://www.jenkins.io/doc/developer/plugin-development/choosing-jenkins-baseline/ -->
     <jenkins.version>2.426.3</jenkins.version>
     <jgit.version>6.9.0.202403050737-r</jgit.version>
+    <mina.api.version>2.12.1-111.v06de17b_e0ed5</mina.api.version>
```
good point. but yes the only very usable place for such comment is around the bom as this the one which will trigger the removal. 
fixed 8e9493bc

olamy commented on src/test/java/org/jenkinsci/plugins/gitclient/jgit/AuthzTest.java:
```diff
@@ -0,0 +1,161 @@
+        client.add("test*");
+        client.commit("Very useful change");
+        client.push().to(new URIish(repoUrl)).execute();
+        testRepo = Files.createTempDirectory("git-client-test");
```
ec68551a

olamy commented on src/test/java/org/jenkinsci/plugins/gitclient/jgit/AuthzTest.java:
```diff
@@ -0,0 +1,161 @@
+        Path testFile = testRepo.resolve("test.txt");
+        Files.deleteIfExists(testFile);
+        Files.createFile(testFile);
+        Files.write(testFile, "Hello".getBytes(StandardCharsets.UTF_8));
```
9ce4dbba

olamy commented on src/test/java/org/jenkinsci/plugins/gitclient/verifier/AcceptFirstConnectionVerifierTest.java:
```diff
@@ -26,12 +32,27 @@ public class AcceptFirstConnectionVerifierTest {
+    @AfterClass
+    public static void unsetHostKeyAlgo() {
+        System.clearProperty(JGitAPIImpl.SSH_CONFIG_PATH);
+    }
```
4ac141bf
hehe something else which take more time when migrating to junit 5 :P

olamy commented on pom.xml:
```diff
@@ -65,9 +65,7 @@
     <jenkins.version>2.426.3</jenkins.version>
-    <jgit.version>6.9.0.202403050737-r</jgit.version>
-    <!-- TODO remove when bom version contains this or higher -->
-    <mina.api.version>2.12.1-111.v06de17b_e0ed5</mina.api.version>
```
ditto

olamy replied:
fixed with 445c10c9

olamy commented on src/test/java/org/jenkinsci/plugins/gitclient/verifier/AcceptFirstConnectionVerifierTest.java:
```diff
@@ -47,38 +68,77 @@ public void testVerifyServerHostKeyWhenFirstConnection() throws Exception {
     }
 
     @Test
+    @Ignore("cannot verify a non hash host while we store hash host")
```
test restored left over from initial work

olamy replied:
@MarkEWaite do yo have some logs? because it's passing locally

MarkEWaite replied:
I've uploaded a gist with the log that happens on my local Linux computer when I run:

```
  mvn clean -Dtest=AcceptFirstConnectionVerifierTest verify 2>&1 | tee logfile
```

I believe that the same message was reported on ci.jenkins.io , though a failing run is not available in the job history of that machine

MarkEWaite replied:
I've pushed 897001ec71e14808cb0e3bca73a2fa0eb450739d so that I can double check if the failure on ci.jenkins.io is the same as the failure that I see locally.  That will help me check if my local configuration is the cause of the local failure that I see.

MarkEWaite replied:
The [ci.jenkins.io **job](https://ci.jenkins.io/job/Plugins/job/git-client-plugin/view/change-requests/job/PR-1127/77/testReport/junit/org.jenkinsci.plugins.gitclient.verifier/AcceptFirstConnectionVerifierTest/linux_21___Build__linux_21____testVerifyServerHostKeyWhenHostnameWithoutPort/)** fails the test with the same message.  It may still be a configuration error, but if so, then the configuration error is both in my environment and on ci.jenkins.io

olamy commented on src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java:
```diff
@@ -163,16 +182,76 @@ public class JGitAPIImpl extends LegacyCompatibleGitAPIImpl {
+    public SshdSessionFactory buildSshdSessionFactory(@NonNull final HostKeyVerifierFactory hostKeyVerifierFactory) {
+        if (Files.notExists(hostKeyVerifierFactory.getKnownHostsFile().toPath())) {
+            try {
+                Files.createDirectories(hostKeyVerifierFactory
+                        .getKnownHostsFile()
+                        .getParentFile()
+                        .toPath());
+                Files.createFile(hostKeyVerifierFactory.getKnownHostsFile().toPath());
```
non secured permissions. not sure what is sensitive in a `known_hosts` file.
if there is default umask with write permission, there are probably even more issues than a  `known_hosts` file :)
btw this has been the case for ages now with the trilead API. not sure why we should changed that.

jtnord replied:
nothing in the `known_hosts` is sensitive, however the `.ssh` directory is and if you create this insecurely then other commands may fail later (ie it could break the build, or leave an agent unable to be used later due to this).

olamy replied:
I know this, but the problem is to change a behavior that has existed for a long time with the current code.
if everybody agrees I don't really mind but it's a change of the result on the files created

olamy replied:
`man ssh` says 
```
     ~/.ssh/
             This directory is the default location for all user-specific configuration and authentication information.  There is no general requirement to keep the en‐
             tire contents of this directory secret, but the recommended permissions are read/write/execute for the user, and not accessible by others.
```

https://linux.die.net/man/1/ssh

jtnord replied:
> `man ssh` says
> 
> ```
>      ~/.ssh/
>              This directory is the default location for all user-specific configuration and authentication information.  There is no general requirement to keep the en‐
>              tire contents of this directory secret, but the recommended permissions are read/write/execute for the user, and not accessible by others.
> ```
> 
> https://linux.die.net/man/1/ssh

and yet the source code does checking...
https://github.com/openssh/openssh-portable/blob/a8fbe2f7d0d96d299ee8e69769e3b51067978748/regress/check-perm.c#L129-L135

regardless, this is going against the recommendations you point out, and this should not be doing that unless there are valid reasons.
If this was an existing behaviour then it does not need to be changed here - I filed https://issues.jenkins.io/browse/JENKINS-73305 for tracking.

olamy replied:
`there is no general requirement... but recommendation` :)
Not sure when this check has been added but I have this locally and `git` works fine
```
olamy@pop-os:~$ ls -alrt ~/ | grep .ssh
drwxrwxr-x  2 olamy olamy     4096 Apr 29 16:01 .ssh_backup
drwxrwxr-x  2 olamy olamy     4096 Jun 13 18:30 .ssh
-rw-------  1 olamy olamy      339 Jun 13 18:32 .lesshst
olamy@pop-os:~$ ls -alrt ~/.ssh/
total 324
-rw-------  1 olamy olamy    381 Jun  1  2023 id_rsa.pub
-rw-------  1 olamy olamy   1675 Jun  1  2023 id_rsa
-rw-------  1 olamy olamy    611 Jun  1  2023 id_dsa.pub
-rw-------  1 olamy olamy    609 Jun  1  2023 id_dsa_apache.pub
-rw-------  1 olamy olamy    736 Jun  1  2023 id_dsa_apache
-rw-------  1 olamy olamy    668 Jun  1  2023 id_dsa
-rw-------  1 olamy olamy    387 Jun  1  2023 github_rsa.pub
-rw-------  1 olamy olamy   1766 Jun  1  2023 github_rsa
-rw-------  1 olamy olamy   1226 Jun  1  2023 authorized_keys
-rw-------  1 olamy olamy    566 Apr 29 16:01 id_rsa_git_client.pub
-rw-------  1 olamy olamy   2602 Apr 29 16:01 id_rsa_git_client
-rw-------  1 olamy olamy    319 Jun 10 12:16 config
-rw-rw-r--  1 olamy olamy 274069 Jun 11 13:39 known_hosts
drwxrwxr-x  2 olamy olamy   4096 Jun 13 18:30 .
drwxr-x--- 71 olamy olamy   4096 Jun 13 18:32 ..
olamy@pop-os:~$ ssh -V
OpenSSH_8.9p1 Ubuntu-3ubuntu0.7, OpenSSL 3.0.2 15 Mar 2022
```

olamy commented on src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java:
```diff
@@ -163,16 +182,65 @@ public class JGitAPIImpl extends LegacyCompatibleGitAPIImpl {
+                    }
+                    return Collections.emptyList();
+                })
+                .setPreferredAuthentications("publickey,password");
```
nope something everything is done via `SmartCredentialsProvider extends CredentialsProvider`

olamy commented on src/main/java/org/jenkinsci/plugins/gitclient/jgit/CredentialsProviderImpl.java:
```diff
@@ -14,26 +14,32 @@
     /**
      * Constructor for CredentialsProviderImpl.
-     *
+     * @deprecated
```
well why not it''s just telling to use the only constructor not deprecated...

MarkEWaite commented on pom.xml:
```diff
@@ -66,6 +66,8 @@
     <!-- https://www.jenkins.io/doc/developer/plugin-development/choosing-jenkins-baseline/ -->
     <jenkins.version>2.426.3</jenkins.version>
     <jgit.version>6.10.0.202406032230-r</jgit.version>
+    <!-- TODO remove when bom version contains this or higher -->
+    <mina.api.version>2.12.1-111.v06de17b_e0ed5</mina.api.version>
```
Plugin bom now includes Apache Mina ssh 2.12.1-113.v4d3ea_5eb_7f72

https://github.com/jenkinsci/bom/pull/3280

https://github.com/jenkinsci/bom/releases/tag/3120.v4d898e1e9fc4
```suggestion
```

MarkEWaite commented on pom.xml:
```diff
@@ -97,11 +99,13 @@
       <!-- Use mina api common plugin rather than JGit transitive dependency inclusion of Apache Mina sshd common -->
       <groupId>io.jenkins.plugins.mina-sshd-api</groupId>
       <artifactId>mina-sshd-api-common</artifactId>
+      <version>${mina.api.version}</version>
```
Plugin bom now includes Apache Mina ssh 2.12.1-113.v4d3ea_5eb_7f72

https://github.com/jenkinsci/bom/pull/3280

https://github.com/jenkinsci/bom/releases/tag/3120.v4d898e1e9fc4
```suggestion
```

MarkEWaite commented on pom.xml:
```diff
@@ -97,11 +99,13 @@
       <!-- Use mina api core plugin rather than JGit transitive dependency inclusion of Apache Mina sshd core -->
       <groupId>io.jenkins.plugins.mina-sshd-api</groupId>
       <artifactId>mina-sshd-api-core</artifactId>
+      <version>${mina.api.version}</version>
```
Plugin bom now includes Apache Mina ssh 2.12.1-113.v4d3ea_5eb_7f72

https://github.com/jenkinsci/bom/pull/3280

https://github.com/jenkinsci/bom/releases/tag/3120.v4d898e1e9fc4
```suggestion
```

MarkEWaite commented on pom.xml:
```diff
@@ -208,15 +212,29 @@
-      <artifactId>trilead-api</artifactId>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
+      <version>2.0.9</version>
```
@olamy I would have applied this change myself directly to your branch, but was not sure if that is a working style that is OK with you.

olamy replied:
Feel free to push the branch as long as the build is green :)

MarkEWaite commented on pom.xml:
```diff
@@ -234,10 +247,27 @@
+          <artifactId>*</artifactId>
+        </exclusion>
+      </exclusions>
+    </dependency>
```
I agree with @jtnord that I don't see the benefit of the test dependency on slf4j-simple.
```suggestion
```

olamy replied:
because it's so simple to configure a file `simplelogger.properties` when some logs are needed.

MarkEWaite commented on pom.xml:
```diff
@@ -262,6 +292,18 @@
+      <artifactId>slf4j-simple</artifactId>
+      <version>2.0.9</version>
+      <scope>test</scope>
+    </dependency>
```
I've confirmed that tests pass locally without this test dependency.
```suggestion
```

MarkEWaite commented on pom.xml:
```diff
@@ -208,15 +208,28 @@
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
+      <scope>provided</scope>
     </dependency>
```
I think we do not need to declare this dependency.  Tests pass without it.
```suggestion
```

MarkEWaite commented on pom.xml:
```diff
@@ -208,15 +208,28 @@
       <artifactId>structs</artifactId>
     </dependency>
     <dependency>
```
No need for the slf4j dependency as far as I can see.
```suggestion
```

MarkEWaite commented:
I've installed an incremental build of this pull request in my test environment and run my hosting providers tests.  There are several new failure cases compared to the released 4.7.0 version of the plugin.  Unfortunately, I've run out of time for today and need to do other things.

I hope to capture the details of the new failures and will post them here when I have confirmed that there is a behavior change between the released code and the pull request.  Sometimes the failures are entirely the fault of my test infrastructure.

olamy commented on pom.xml:
```diff
@@ -234,10 +251,27 @@
+          <artifactId>*</artifactId>
+        </exclusion>
+      </exclusions>
+    </dependency>
```
because to have debug log for something such `org.apache.sshd.` it's so easier to have a simple `simplelogger.properties` file to use.
but it looks this has been removed.

jtnord replied:
in your test code -> `new LoggerRule().capute("org.apache.sshd", Level.FINEST)`
you can also set a default log level for something for `hpi:run`  via [`loggers`](https://jenkinsci.github.io/maven-hpi-plugin/run-mojo.html#loggers)

MarkEWaite commented on src/main/java/org/jenkinsci/plugins/gitclient/verifier/ManuallyProvidedKeyVerifier.java:
```diff
@@ -34,53 +34,55 @@ public AbstractCliGitHostKeyVerifier forCliGit(TaskListener listener) {
             if (File.pathSeparatorChar
                     == ';') { // check whether on Windows or not without sending Functions over remoting
                 // no escaping for windows because created temp file can't contain spaces
```
When a temporary file contains spaces on Windows, it creates problems for command line git, independent of any issues for the git client plugin.  A [warning is written to the build log](https://github.com/jenkinsci/git-client-plugin/blob/9ca1f0f85c80dd3c20626c1f3a516719f4ed496e/src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java#L240) if a Windows temporary directory is detected with spaces because of the known issue in some versions of command line git.

MarkEWaite commented on src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java:
```diff
@@ -163,16 +182,76 @@ public class JGitAPIImpl extends LegacyCompatibleGitAPIImpl {
+                    try {
+                        String keys = String.join(System.lineSeparator(), sshUserPrivateKey.getPrivateKeys());
+                        return SecurityUtils.loadKeyPairIdentities(
+                                null, () -> "key", IOUtils.toInputStream(keys, StandardCharsets.UTF_8), null);
```
The interactive failures that I saw yesterday in my hosting provider tests were all for JGit implementations using an ssh private key with a passphrase.  I'd like to look into this further today if time allows.  That will help me better understand the code changes.  @olamy no problem if you investigate it as well, but this seems like a good place for me to learn more about the changes.

olamy replied:
should be fixed with this 3ee4346c
sadly the unit test with testcontainer is not really restrictive and not fail if passphrase not provided.

MarkEWaite replied:
I've confirmed that it is fixed in [my installation](https://github.com/MarkEWaite/docker-lfs/blob/4a5e367185031e3c447e40c0635d95a51d07a5bf/plugins.txt#L55).  Thanks making that change!

The JGit credentials with passphrase that were previously failing in my lab environment are now working well with Assembla, Beanstalk, Bitbucket, Gitlab, GitHub, and Microsoft.

olamy commented on src/main/java/org/jenkinsci/plugins/gitclient/verifier/AbstractJGitHostKeyVerifier.java:
```diff
@@ -1,64 +1,115 @@
 
-    KnownHosts getKnownHosts() {
-        return knownHosts;
+    protected boolean askAboutKnowHostFile() {
```
ba943972

olamy commented:
@MarkEWaite I managed to make the test failure repeatable (a bit flaky to async io aspect of mina)
just need to figure out the problem now.

olamy commented on src/test/java/org/jenkinsci/plugins/gitclient/verifier/AcceptFirstConnectionVerifierTest.java:
```diff
@@ -130,7 +130,7 @@ public void testVerifyServerHostKeyWhenHostnameWithoutPort() throws Exception {
         }
         String hostKeyEntry =
-                "github.com ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBEmKSENjQEezOmxkZMy7opKgwFB9nkt5YRrYMjNuG5N87uRgg6CLrbo5wAdT/y6v0mKV0U2w0WZ2YB/++Tpockg=]";
+                "github.com ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBEmKSENjQEezOmxkZMy7opKgwFB9nkt5YRrYMjNuG5N87uRgg6CLrbo5wAdT/y6v0mKV0U2w0WZ2YB/++Tpockg=";
```
@MarkEWaite  LOL I just  wasted few hours only because of a ridiculous typo but couldn't see it :rofl:

olamy replied:
@MarkEWaite haha no worries. I'm just complaining about myself not seeing this quicker. 
two pairs of eyes are not enough

olamy commented on src/test/java/org/jenkinsci/plugins/gitclient/verifier/AcceptFirstConnectionVerifierTest.java:
```diff
@@ -47,38 +68,77 @@ public void testVerifyServerHostKeyWhenFirstConnection() throws Exception {
+        // AAAAC3NzaC1lZDI1NTE5AAAAIOMqqnkVzrm0SdG6UOoqKLsabgH5C9okWi0dh2l9GKJl";
         String hostKeyEntry =
-                "|1|I9eFW1PcZ6UvKPt6iHmYwTXTo54=|PyasyFX5Az4w9co6JTn7rHkeFII= ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOMqqnkVzrm0SdG6UOoqKLsabgH5C9okWi0dh2l9GKJl";
+                "|1|FJGXVAi7jMQIsl1J6uE6KnCiteM=|xlH92KQ91GuBgRxvRbU/sBo60Bo= ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBEmKSENjQEezOmxkZMy7opKgwFB9nkt5YRrYMjNuG5N87uRgg6CLrbo5wAdT/y6v0mKV0U2w0WZ2YB/++Tpockg=";
```
https://github.com/apache/mina-sshd/blob/d45646b08247e1578b367ed8cf2c326ebd22cd52/sshd-common/src/main/java/org/apache/sshd/common/signature/BuiltinSignatures.java#L155

olamy commented on src/main/java/org/jenkinsci/plugins/gitclient/verifier/KnownHostsFileVerifier.java:
```diff
@@ -32,51 +31,35 @@ public AbstractCliGitHostKeyVerifier forCliGit(TaskListener listener) {
-                knownHosts = new KnownHosts();
+                Path parent = knowHostPath.getParent();
+                if (parent != null) {
+                    Files.createDirectories(parent);
```
same response as here :) https://github.com/jenkinsci/git-client-plugin/pull/1127/files#r1632002556

MarkEWaite commented on src/main/java/org/jenkinsci/plugins/gitclient/jgit/SmartCredentialsProvider.java:
```diff
@@ -1,12 +1,17 @@
-package org.jenkinsci.plugins.gitclient.trilead;
+package org.jenkinsci.plugins.gitclient.jgit;
```
Won't this change break runtime and compilation for the plugins that are using the import, like:

* [Blue ocean plugin](https://github.com/jenkinsci/blueocean-plugin/blob/6d6e0beef662fdcdca57459ed4c24733a2e4fac4/blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitUtils.java#L63)
* [Gerrit code review plugin](https://github.com/jenkinsci/gerrit-code-review-plugin/blob/c53550150be123d96751f2f5d5796e868c7be43c/src/main/java/jenkins/plugins/gerrit/UsernamePasswordCredentialsProvider.java#L20)

Do we need to retain a class at `org.jenkinsci.plugins.gitclient.trilead.SmartCredentialsProvider` that calls the new implementation at `org.jenkinsci.plugins.gitclient.jgit.SmartCredentialsProvider`?

olamy replied:
oh no.... we have to live with this first bad choice of naming :(

olamy replied:
fixed with 1b4531ff

olamy replied:
oh no BlueOcean :man_facepalming: 
just looking at it and I used to be the last man standing with the only laptop being able to build BO locally.
But it looks this time have gone and I get `npm` failing with a `segfault` :) 
bear with me I will look at  this.

olamy replied:
@MarkEWaite sadly not much I can do except adapt it for next version.
I cannot change error message from Apache Mina (see comment here https://github.com/jenkinsci/blueocean-plugin/pull/2561/files#diff-86625d110303de518f61fe53504435aacddf203575cf17b678d75988e65d4c06R141)
Anyway it's definitely not blocker

MarkEWaite replied:
That sounds reasonable to me.  Are you OK with releasing a new version of Blue Ocean in the same week that the new version of git client plugin is released?

If not, we can also exclude those test failures on the plugin bill of materials, but then we need to remember to remove the exclusions after the next release of the plugin BOM.

olamy replied:
I can even release BlueOcean now as the code change will work for current git client version as well

olamy replied:
BO release done.

MarkEWaite replied:
Confirmed that the blueocean git plugin test in the plugin BOM passes.  Thanks!

I've merged the master branch to this branch and plan to submit the incremental to the bom repository so that it can check these changes and JGit 6.10.0 with the other plugins in the BOM.

MarkEWaite replied:
https://github.com/jenkinsci/bom/pull/3299 is the plugin BOM test that is checking the most recent build.

MarkEWaite replied:
https://github.com/jenkinsci/bom/pull/3299 passed all but 3 tests.  Those 3 tests failed with an artifact caching proxy timeout.  The 3 tests pass locally for me.  I consider that a "go" for the plugin bill of materials.

MarkEWaite commented:
During some testing I found that the known hosts file stored in the temporary directory was corrupted or damaged on my one Windows agent that has allows more than one executor.  I'm too tired to continue the testing, but I am suspicious enough of the thread safety of the known hosts file handling that I'll do some further testing the next time I have a chance.  That won't be tomorrow because tomorrow is filled with other things.

I was running all my hosting provider tests at the same time and queuing multiple runs.  Most of my Windows agents only have a single executor assigned, but one machine has 3 executors.  That machine is the one that scrambled the contents of the known hosts file.

olamy commented:
> During some testing I found that the known hosts file stored in the temporary directory was corrupted or damaged on my one Windows agent that allows more than one executor. I'm too tired to continue the testing, but I am suspicious enough of the thread safety of the known hosts file handling that I'll do some further testing the next time I have a chance. That won't be tomorrow because tomorrow is filled with other things.
> 
> I was running all my hosting provider tests at the same time and queuing multiple runs. Most of my Windows agents only have a single executor assigned, but one machine has 3 executors. That machine is the one that scrambled the contents of the known hosts file.

I will need more information such: 
- what verifier strategy is used?
- did you start with an empty known hosts file
- any error in jenkins log or node logs?

This would be surprising to see concurrent access as this should be really better now with JGit as it is using a LockFile implementation (see https://github.com/eclipse-jgit/jgit/blob/master/org.eclipse.jgit/src/org/eclipse/jgit/internal/storage/file/LockFile.java) compared to trilead which was using simply no sync mechanism (see https://github.com/jenkinsci/trilead-ssh2/blob/15dfef32ff14a4c3eba66552f3636f886ad340c4/src/com/trilead/ssh2/KnownHosts.java#L648)

MarkEWaite commented:
Thanks for asking the questions.  I haven't investigated enough to give really good answers yet, but here are some of the things that I can answer.

> I will need more information such:
> 
> * what verifier strategy is used?

I'm using manually provided key verification strategy.  Details are in the [configuration as code of my controller](https://github.com/MarkEWaite/docker-lfs/blob/9daf97bce8394dd04b954863a56bdf856fa990a6/ref/configuration-as-code/security.yaml#L16).

> * did you start with an empty known hosts file

I believe that I started with a non-existent known hosts file because the one agent log reported that a temporary file that included the word "knownhosts" in its name did not contain the host key for bitbucket.org.  Since it was a file in a temporary directory with "knownhosts" and a number in the file name, I assume it was created by the agent some time after the agent started.  

Unfortunately, I don't have the exact message but I believe it was related to the data in the known hosts file being somehow incorrect or incomplete.

> * any error in jenkins log or node logs?

Yes, there were multiple messages in the agent log that indicated the temporary file with "knownhosts" in its name did not contain the host key for several different Git providers.  I believe that Bitbucket and Gitea were both mentioned and I think that one other was mentioned.

I'll experiment further with it on my next testing round.

olamy commented:
> 
> That script starts 96 jobs, with at least a few of the jobs as Pipeline jobs that perform checkout on multiple agents. It immediately loads my 47 agents with jobs. When I run that script a second time, it doubles the number of jobs.

does this work with current code (`master`) using trilead implementation?

TBH the goal of this PR is to change from Trilead to Apache Mina.
Same for @jtnord comment of directory permissions
Do we really want to add out of context code change? 
I usually get comments on non-related changes (such as extra lines or spaces added somewhere), and now you are asking about behavior changes. so I'm lost 😆 
At this stage, this PR has reached a status where it works (including bug backward compatibility :)).
I don't mind adding extra nonrelated changes, but personally, I would prefer having this merged and released to get some more feedback early, as there will obviously be some bugs. However, we cannot retain this for ages and delay all the next feedbacks from more users with different configuration setups, etc.

olamy commented:
@MarkEWaite, I looked again at the manually provided verification strategy. It's using his own file per client instance (same implementation as `master` with trilead or even the cli one, so there shouldn't be any concurrent access issues.

olamy commented:
@MarkEWaite do you have other plugins installed such https://github.com/jenkinsci/eddsa-api-plugin?

MarkEWaite commented:
> does this work with current code (`master`) using trilead implementation?

An excellent question.  I should have thought of that myself.  Further exploration shows that my initial theory of a concurrency issue was wrong.  I added extra executors to all of the Windows machines in my environment, some with processors that are faster than the problematic machine, some with processors that are slower.  None of the other machines showed any issue.  I think the problem I was seeing is specific to that machine and is not worth delaying this pull request.

> Do we really want to add out of context code changes?

No, I definitely don't want to add code changes that are out of context.  You've done great work and my testing confirms that great work.

I think we should squash merge this change and release immediately.  Does that work for you?

MarkEWaite commented:
> @MarkEWaite do you have other plugins installed such https://github.com/jenkinsci/eddsa-api-plugin?

Yes, I have eddsa plugin installed in [my plugins.txt](https://github.com/MarkEWaite/docker-lfs/blob/78616195e2ce8c3df7cd27e47f34eed4baa8a378/plugins.txt#L48).  I believe that it is installed as a dependency of ssh server plugin and trilead API plugin.

olamy commented:
> > does this work with current code (`master`) using trilead implementation?
> 
> An excellent question. I should have thought of that myself. Further exploration shows that my initial theory of a concurrency issue was wrong. I added extra executors to all of the Windows machines in my environment, some with processors that are faster than the problematic machine, some with processors that are slower. None of the other machines showed any issue. I think the problem I was seeing is specific to that machine and is not worth delaying this pull request.
> 
> > Do we really want to add out of context code changes?
> 
> No, I definitely don't want to add code changes that are out of context. You've done great work and my testing confirms that great work.
> 
> I think we should squash merge this change and release immediately. Does that work for you?

I just fixed some merge conflict with `master`.
Sounds definitely good to squash/merge -> release -> celebrate (until the first user's bug :) )

MarkEWaite commented:
> > I think we should squash merge this change and release immediately. Does that work for you?
> 
> I just fixed some merge conflict with `master`. Sounds definitely good to squash/merge -> release -> celebrate (until the first user's bug :) )

I think we should increment the version number to 5.0.0 because this is intentionally deleting some public classes.  I've checked that they are unused public classes, but I think that consumers will be better served by knowing that the public API has intentionally changed to remove many of the trilead classes.  I'll do that in a separate pull request after this is merged, then will run a quick test to confirm.  I hope to have that completed before the weekend is done.

MarkEWaite commented:
> I just fixed some merge conflict with `master`.

There wasn't a merge conflict with master as far as I can tell.  The pull request had 8176eab5 as a merge commit that brought in the most recent master branch.  It looks like your merge resolution added many more commits to the history.

olamy commented:
> > I just fixed some merge conflict with `master`.
> 
> There wasn't a merge conflict with master as far as I can tell. The pull request had [8176eab](https://github.com/jenkinsci/git-client-plugin/commit/8176eab58d22f9e613129c41df7f0ee9cd744c63) as a merge commit that brought in the most recent master branch. It looks like your merge resolution added many more commits to the history.

glursp that's weird. I just used a standard `git rebase origin/master` 
I bumper version to 5.0.0
anyway what do you prefer squashing when you merge via UI or I can do it locally.
I don't really have strong preference at the end of the day it's the same result.

MarkEWaite commented:
> anyway what do you prefer squashing when you merge via UI or I can do it locally.

If you're willing to do the squash merge, that's great for me.  That allows you to have local control of it and be confident of the content of the squash.

MarkEWaite commented:
It has been a long day for me.  I'll let you complete the squash and the merge and I'll plan to do more testing tomorrow.  I need some sleep now.

olamy commented:
> It has been a long day for me. I'll let you complete the squash and the merge and I'll plan to do more testing tomorrow. I need some sleep now.

squash done.
no worries. have some rest.  Thanks for your help.
